### ISSUE TITLE
Will getContactPoints return infinite points?

### ISSUE THREAD

**Issue by jingxixu** [bulletphysics/bullet3#4221]
Hi everyone, I am trying to understand the function [getContactPoints](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.cb0co8y2vuvc). If two solid bodies (meshes) penetrate, there should be an infinite number of contact points. **Imagine two triangular faces intersect at a line and the line will contain infinite number of points.** How does this function handle it so that it does not return infinite number of contact points? Thanks a lot!

**Comment by xissburg** [comment]
There's a limit of four contact points, which define a contact polygon. As part of collision detection, there is an algorithm which chooses which points to keep as to maximize the area of the contact polygon.

**Comment by jingxixu** [comment]
Just to confirm, is the contact polygon the intersection surface between the two rigid bodies, as shown in the figure below? Do you mind sharing a reference to the code/algorithm that you talked about? Many thanks!

![Image from iOS (12)](https://user-images.githubusercontent.com/15113241/160430943-fcab1ca4-7fa4-4437-9596-c28ccff1bdd8.jpg)

**Comment by xissburg** [comment]
Yes, pretty much. In your example, there should be 4 contact points, one at each vertex of the dashed rectangle. 
The [btPersistentManifold::addManifoldPoint](https://github.com/bulletphysics/bullet3/blob/22d9eec148d3a7dec9d7b10a4344808bf766f73a/src/BulletCollision/NarrowPhaseCollision/btPersistentManifold.cpp#L212) function will select which point to replace if the maximum count has been reached. `btPersistentManifold::sortCachedPoints` implements that logic using the `calcArea4Points` function to make the best choice..
